criu/pages-compress.c

```diff
@@ -0,0 +1,285 @@
+#include <limits.h>
+#include <pthread.h>
+#include <sched.h>
+#include <stdio.h>
+#include <sys/mman.h>
+#include <sys/resource.h>
+#include <unistd.h>
+
+#include <compel/plugins/std/syscall-codes.h>
+
+#include "pages-compress.h"
+
+#include "crtools.h"
+#include "image.h"
+#include "log.h"
+#include "memfd.h"
+#include "sched.h"
+#include "util.h"
+
+#include <lz4/programs/lz4io.h>
+#include <lz4/lib/lz4frame.h>
+
+#undef LOG_PREFIX
+#define LOG_PREFIX "decompress: "
+
+const char * const decompressed_filename_format = "/tmp/decompressed_pid%d.img";
+static char decompressed_filename[PATH_MAX] = {0};
+
+// Decompression result: 0 - success; negative - decompression error occured.
+static int decompressionResult = -ENOENT;
+
+int compress_images(void)
+{
+	char format[PATH_MAX];
+	char pathsrc[PATH_MAX];
+	char pathdst[PATH_MAX];
+	const long pid = getpid();
+	const int dfd = get_service_fd(IMG_FD_OFF);
+	const int id = 1; // the only image we'd like to compress
+	int ret;
+
+	LZ4IO_prefs_t *lz4_prefs = LZ4IO_defaultPreferences();
+	LZ4IO_setBlockSize(lz4_prefs, 128 * 1024);
+
+	pr_debug("Compressing pages\n");
+
+	snprintf(format, sizeof(format), "%s/%s", "/proc/%ld/fd/%d", imgset_template[CR_FD_PAGES].fmt);
+	snprintf(pathsrc, sizeof(pathsrc), format, pid, dfd, id);
+
+	snprintf(format, sizeof(format), "%s/%s", "/proc/%ld/fd/%d", imgset_template[CR_FD_PAGES_COMP].fmt);
+	snprintf(pathdst, sizeof(pathdst), format, pid, dfd, id);
+
+	ret = LZ4IO_compressFilename(pathsrc, pathdst, 8, lz4_prefs);
+	LZ4IO_freePreferences(lz4_prefs);
+
+	if (ret) {
+		pr_err("Can't compress %s to %s\n", pathsrc, pathdst);
+		return -1;
+	}
+	if (unlink(pathsrc)) {
+		pr_perror("Can't delete uncompressed source pages image: %s\n", pathsrc);
+		return -1;
+	}
+	return 0;
+}
+
+static int decompress_image3(const char *mapped_addr, size_t mapped_size, LZ4F_decompressionContext_t dctx) {
+	const char *compbuf = mapped_addr;
+	const int outbufsize = 4 * 1024 * 1024; // LZ4F_max4MB is the max block size; see LZ4F_blockSizeID_t for details
+	char outbuf[outbufsize];
+	LZ4F_errorCode_t lz4err;
+	int decomp_fd;
+	size_t totalread = 0;
+	size_t totalwrite = 0;
+
+	snprintf(decompressed_filename, sizeof(decompressed_filename), decompressed_filename_format, (int)getpid());
+	pr_debug("Creating tmp file at '%s'\n", decompressed_filename);
+	decomp_fd = open(decompressed_filename, O_CREAT | O_WRONLY | O_TRUNC, S_IRUSR | S_IWUSR);
+
+	if (0 > decomp_fd) {
+		pr_err("Can't create file, errno=%d\n", errno);
+		return -1;
+	}
+
+	while (1) {
+		size_t insize = mapped_size - (compbuf - mapped_addr);
+		if (!insize) {
+			/* reached end of file or stream, no data left to decompress */
+			decompressionResult = 0;
```

Comment:
Aren't return value of these functions now pointless? Only `decompressionResult` matters.

Reply:
Yes, but I'd like to make functions consistent returning a value, rather than checking a global variable every time. This may help for further refactoring. Currently the implementation checks the global variable once, in the place it really needs.